circuit/src/gadgets/main.rs

```diff
@@ -0,0 +1,950 @@
+		// | x   | y   | res |     |    |
+
+		let (zero, product) = layouter.assign_region(
+			|| "assign_values",
+			|region| {
+				let mut ctx = RegionCtx::new(region, 0);
+				let zero = ctx.assign_advice(common.advice[0], Value::known(F::zero()))?;
+				let product =
+					ctx.assign_advice(common.advice[1], self.x.value().cloned() * self.y.value())?;
+				Ok((zero, product))
+			},
+		)?;
+
+		let advices = [self.x, self.y, product.clone(), zero.clone(), zero];
+		let fixed =
+			[F::zero(), F::zero(), -F::one(), F::zero(), F::zero(), F::one(), F::zero(), F::zero()];
+		let main_chip = MainChip::new(advices, fixed);
+		main_chip.synthesize(common, &config.selector, layouter)?;
+
+		Ok(product)
+	}
+}
+
+/// Chip for is_bool operation
+pub struct IsBoolChipset<F: FieldExt> {
+	x: AssignedCell<F, F>,
+}
+
+impl<F: FieldExt> IsBoolChipset<F> {
+	/// Create new IsBoolChipset
+	pub fn new(x: AssignedCell<F, F>) -> Self {
+		Self { x }
+	}
+}
+
+impl<F: FieldExt> Chipset<F> for IsBoolChipset<F> {
+	type Config = MainConfig;
+	type Output = ();
+
+	fn synthesize(
+		self, common: &crate::CommonConfig, config: &Self::Config, mut layouter: impl Layouter<F>,
+	) -> Result<Self::Output, Error> {
+		// We should satisfy the equation below
+		// (1 - x) * x = 0
+		// x - x * x = 0
+
+		// Witness layout:
+		// | A   | B   | C   | D   | E  |
+		// | --- | --- | --- | --- | ---|
+		// | x   |     | x   | x   |    |
+
+		let zero = layouter.assign_region(
+			|| "assign_values",
+			|region| {
+				let mut ctx = RegionCtx::new(region, 0);
+				let zero = ctx.assign_advice(common.advice[0], Value::known(F::zero()))?;
+				Ok(zero)
+			},
+		)?;
+
+		// [a, b, c, d, e]
+		let advices = [self.x.clone(), zero.clone(), self.x.clone(), self.x.clone(), zero];
+		// [s_a, s_b, s_c, s_d, s_e, s_mul_ab, s_mul_cd, s_constant]
+		let fixed =
+			[F::one(), F::zero(), F::zero(), F::zero(), F::zero(), F::zero(), -F::one(), F::zero()];
+		let main_chip = MainChip::new(advices, fixed);
+		main_chip.synthesize(common, &config.selector, layouter)?;
+
+		Ok(())
+	}
+}
+
+/// Chip for is_equal operation
+pub struct IsEqualChipset<F: FieldExt> {
+	x: AssignedCell<F, F>,
+	y: AssignedCell<F, F>,
+}
+
+impl<F: FieldExt> IsEqualChipset<F> {
+	/// Create new IsEqualChipset
+	pub fn new(x: AssignedCell<F, F>, y: AssignedCell<F, F>) -> Self {
+		Self { x, y }
+	}
+}
+
+impl<F: FieldExt> Chipset<F> for IsEqualChipset<F> {
+	type Config = MainConfig;
+	type Output = AssignedCell<F, F>;
+
+	fn synthesize(
+		self, common: &crate::CommonConfig, config: &Self::Config, mut layouter: impl Layouter<F>,
+	) -> Result<Self::Output, Error> {
+		// We should satisfy the equation below
+		// x - y = 0
+
+		let diff = layouter.assign_region(
+			|| "assign_values",
+			|region| {
+				let diff = self.x.clone().value().cloned() - self.y.value();
+
+				let mut ctx = RegionCtx::new(region, 0);
+				let diff = ctx.assign_advice(common.advice[0], diff)?;
+				Ok(diff)
+			},
+		)?;
```

Comment:
You are calculating the difference between `x` and `y`, but no constrains are actually applied.
You probably need substraction chipset in here

Reply:
Done in https://github.com/eigen-trust/protocol/pull/132/commits/fd67e89d0f674fd64a7f6a4f38e98e67b60db69f